### PR TITLE
(PDK-1438) Auto select Facter version based on Puppet version

### PR DESCRIPTION
**Pull request by rodjek**
An endpoint has been added to the Forge API allowing us to pull down
a list of puppet-agent AIO package versions and the components inside
them. With this we can now automatically look up the used Puppet version
and select the appropriate Facter version to use when querying FacterDB.

When looking up the Puppet version, it will look backwards through the
known Puppet versions to find the next best match if there is not
a direct match. This means that when a new Puppet version is released,
it will select the Facter version for the latest version of Puppet that
it knows about, resulting in a pretty close match. If it searches
backwards and can not find any Puppet versions then it will default to
the currently loaded Facter version.

### CODE REVIEW

**Comment by rodjek** [comment]
While the coverage % has decreased (due to the addition of lines of code), none of the uncovered lines are related to this change.

**Comment by mcanevet** [comment]
LGTM

**Comment by igalic** [comment]
If i'm reading this code right, then this works offline, and only the rake task updates the `ext/puppet_agent_components.json` file. — if so, perhaps add that to your commit message.

File: lib/rspec-puppet-facts.rb [hunk 1]
```diff
@@ -348,4 +343,35 @@ def self.down_facter_version(version, minor_subtractor = 1)
...
     "#{major}.#{minor}.#{z}"
   end
+
+  def self.facter_version_for_puppet_version(puppet_version)
+    return Facter.version if puppet_version.nil?
+    json_path = File.expand_path(File.join(__dir__, '..', 'ext', 'puppet_agent_components.json'))
+    return Facter.version unless File.file?(json_path) && File.readable?(json_path)
+
+    fd = File.open(json_path, 'rb')
+    data = JSON.parse(fd.read)
```
Comment (scotje) [inline lib/rspec-puppet-facts.rb]: Is it worth rescuing a JSON parse error here?

File: lib/rspec-puppet-facts.rb [hunk 2]
```diff
@@ -348,4 +343,35 @@ def self.down_facter_version(version, minor_subtractor = 1)
...
+
+    version_map = data.map { |_, versions|
+      if versions['puppet'].nil? || versions['facter'].nil?
+        nil
+      else
+        [Gem::Version.new(versions['puppet']), versions['facter']]
+      end
+    }.compact
+
+    gem_version = Gem::Version.new(puppet_version)
```
Comment (scotje) [inline lib/rspec-puppet-facts.rb]:
```suggestion
    puppet_gem_version = Gem::Version.new(puppet_version)
```

Comment (scotje) [reply lib/rspec-puppet-facts.rb]: And next line too but Github doesn't let you do multiline suggestions :/

File: Rakefile [hunk 3]
```diff
@@ -14,3 +16,34 @@ task :dump_commit do
...
+
+    uri = URI.parse('https://forgeapi.puppet.com/private/versions/puppet-agent')
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = uri.scheme == 'https'
+
+    request = Net::HTTP::Get.new(uri.request_uri)
+    response = http.request(request)
+    raise unless response.is_a?(Net::HTTPSuccess)
+
+    File.open(PUPPET_VERSIONS_PATH, 'wb') do |fd|
```
Comment (glennsarti) [inline Rakefile]: nit: use utf8 encoding

File: lib/rspec-puppet-facts.rb [hunk 4]
```diff
@@ -348,4 +343,35 @@ def self.down_facter_version(version, minor_subtractor = 1)
     minor = (minor.to_i - minor_subtractor).to_s
     "#{major}.#{minor}.#{z}"
   end
+
+  def self.facter_version_for_puppet_version(puppet_version)
+    return Facter.version if puppet_version.nil?
+    json_path = File.expand_path(File.join(__dir__, '..', 'ext', 'puppet_agent_components.json'))
+    return Facter.version unless File.file?(json_path) && File.readable?(json_path)
+
+    fd = File.open(json_path, 'rb')
```
Comment (glennsarti) [inline lib/rspec-puppet-facts.rb]: nit: use utf8 encoding

File: spec/rspec_puppet_facts_spec.rb [hunk 5]
```diff
@@ -377,31 +457,31 @@
...
+        it { is_expected.to include('windows-2012 R2-x86_64' => an_instance_of(Hash)) }
       end
 
       context 'with a Server prefixed release' do
         let(:release) { ['Server 2012'] }
 
         it { is_expected.to be_a(Hash) }
         it { is_expected.to have_attributes(:size => 1) }
-        it { is_expected.to include('windows-2012-x64' => an_instance_of(Hash)) }
+        it { is_expected.to include('windows-2012-x86_64' => an_instance_of(Hash)) }
```
Comment (glennsarti) [inline spec/rspec_puppet_facts_spec.rb]: Curious why the platform change here

Comment (rodjek) [reply spec/rspec_puppet_facts_spec.rb]: It looks like the `hardwaremodel` value changed in the Facter 3.x series for Windows from `x64` to `x86_64`.

File: lib/rspec-puppet-facts.rb (hunk 1 above)
Comment (rodjek) [reply lib/rspec-puppet-facts.rb]: :+1:

File: lib/rspec-puppet-facts.rb (hunk 2 above)
Comment (rodjek) [reply lib/rspec-puppet-facts.rb]: :+1:

File: Rakefile (hunk 3 above)
Comment (rodjek) [reply Rakefile]: :+1:

File: lib/rspec-puppet-facts.rb (hunk 4 above)
Comment (rodjek) [reply lib/rspec-puppet-facts.rb]: :+1:

**Pull request merged**